### DISCUSSION TITLE
Questions about python-opcua

### DISCUSSION THREAD

**Discussion by Stequila93** [FreeOpcUa/python-opcua#1483] [General]
Hello,
I'm just using for the first time python and this library.
I made a simple script that connect to more than 1 server, manage connection and reconnection and a coroutine that sometimes periodically will get values from a Siemens PLC.
Seems to be just 10 int variables for each client, is it better to do a subscription? or get_values? at the moment i'm using read_value() from nodes childs.
Also i noticed that if the session is in timeout i cannot use activate_session(), i'm gettin BadSessionIdInvalid.
Does the server close the session and client need to connect again if its get timeout?
Also seems that the PLC has its own timeout around 10 seconds (i didnt watch about settings in tia portal), if i would get data every minute should i write/read a variable every 10 second to keep session alive?
Thanks a lot.

**Comment by swamper123** [comment]
Hey ho,

>  I'm just using for the first time python and this library.

first things first: This library is deprecated #1475 , it would be better to swtich to [opcua-asyncio](https://github.com/FreeOpcUa/opcua-asyncio).

PLCs are getting more and more into OPC-UA and it's mechanisms (like Subscriptions, History etc.).
From an OPC UA point of view, a subscription or Pub-Sub would be the smoothiest, most efficient way. But not every PLC is supporting it right now.
What ~~will~~ should defitnitly work is to request multiple Nodes in one request every x seconds.

Yes, the session timeout after (I think it was around 1hr here as default) time x is wanted behaviour. Normally in such a case a reconnect procedure would happen. AFAIK we still haven't got a clean way (neither here nor in asyncua), just a workaround.

But before we dive in deeper discussions, you may want to switch to opcua-asyncio, since there have been a lot of fixes being made. :)

**Comment by oroulet** [comment]
> IWhat will should defitnitly work is to request multiple Nodes in one request every x seconds.

there is a method to read many nodes, just call it every X seconds

The timeout stuff should not be any issue, you probably have some setup issue on PLC or client side. We are staying connected for months against beckhoff PLCs here.

**Comment by Stequila93** [comment]
Thanks  got it.
I'm already using asyncio library, I misspelled the title.
Already changed the method of reading variables to get_values() thats better.
I'm still not watching opcua server settings on Sie, but i think i'm not gonna need anything, because i just saw that i will need to get data every 5s so the connecttion will be keep alive.

Thanks again.
